File: src/cjhx_agile_workflow/lifecycle.py
from __future__ import annotations

from collections.abc import Iterable

from .errors import TransitionError
from .models import Change, LifecycleState, TransitionRecord, utc_now


ALLOWED_TRANSITIONS: dict[LifecycleState, set[LifecycleState]] = {
    LifecycleState.INTENT_DRAFT: {LifecycleState.INTENT_CONFIRMED},
    LifecycleState.INTENT_CONFIRMED: {LifecycleState.REQUIREMENT_READY, LifecycleState.INTENT_DRAFT},
    LifecycleState.REQUIREMENT_READY: {LifecycleState.DESIGN_APPROVED, LifecycleState.INTENT_CONFIRMED},
    LifecycleState.DESIGN_APPROVED: {LifecycleState.IMPLEMENTING, LifecycleState.REQUIREMENT_READY},
    LifecycleState.IMPLEMENTING: {LifecycleState.REVIEWING, LifecycleState.REQUIREMENT_READY},
    LifecycleState.REVIEWING: {LifecycleState.VERIFIED, LifecycleState.IMPLEMENTING},
    LifecycleState.VERIFIED: {LifecycleState.ACCEPTED, LifecycleState.IMPLEMENTING},
    LifecycleState.ACCEPTED: {LifecycleState.RELEASE_APPROVED, LifecycleState.REQUIREMENT_READY},
    LifecycleState.RELEASE_APPROVED: {LifecycleState.DEPLOYING, LifecycleState.IMPLEMENTING},
    LifecycleState.DEPLOYING: {LifecycleState.OPERATING, LifecycleState.IMPLEMENTING},
    LifecycleState.OPERATING: {LifecycleState.OUTCOME_VALIDATED, LifecycleState.IMPLEMENTING},
    LifecycleState.OUTCOME_VALIDATED: {LifecycleState.REQUIREMENT_READY},
}

REQUIRED_EVIDENCE: dict[LifecycleState, set[str]] = {
    LifecycleState.INTENT_CONFIRMED: {"intent-approval"},
    LifecycleState.REQUIREMENT_READY: {"requirement-spec"},
    LifecycleState.DESIGN_APPROVED: {"technical-design", "design-approval"},
    LifecycleState.REVIEWING: {"change-request"},
    LifecycleState.VERIFIED: {"code-review", "quality-verification"},
    LifecycleState.ACCEPTED: {"acceptance-approval"},
    LifecycleState.RELEASE_APPROVED: {"release-plan", "release-approval"},
    LifecycleState.OPERATING: {"deployment-record"},
    LifecycleState.OUTCOME_VALIDATED: {"outcome-validation"},
}


def missing_evidence(change: Change, target: LifecycleState) -> set[str]:
    present = {item.kind for item in change.evidence if item.status.lower() in {"passed", "approved", "valid"}}
    return REQUIRED_EVIDENCE.get(target, set()) - present


def transition(
    change: Change,
    target: LifecycleState,
    *,
    actor: str,
    reason: str,
    evidence_ids: Iterable[str] = (),
    enforce_gates: bool = True,
) -> Change:
    target = LifecycleState(target)
    if target not in ALLOWED_TRANSITIONS[change.state]:
        raise TransitionError(f"cannot transition from {change.state.value} to {target.value}")
    missing = missing_evidence(change, target) if enforce_gates else set()
    if missing:
        raise TransitionError(f"transition gate is missing evidence: {', '.join(sorted(missing))}")
    previous = change.state
    change.state = target
    change.updated_at = utc_now()
    change.history.append(
        TransitionRecord(
            from_state=previous.value,
            to_state=target.value,
            actor=actor,
            reason=reason,
            evidence_ids=list(evidence_ids),
        )
    )
    return change

File: src/lifecycle.ts
import { TransitionError } from "./errors.js";
import type { Change, LifecycleState } from "./models.js";
import { utcNow } from "./models.js";

const allowed: Record<LifecycleState, Set<LifecycleState>> = {
  intent_draft: new Set(["intent_confirmed"]),
  intent_confirmed: new Set(["requirement_ready", "intent_draft"]),
  requirement_ready: new Set(["design_approved", "intent_confirmed"]),
  design_approved: new Set(["implementing", "requirement_ready"]),
  implementing: new Set(["reviewing", "requirement_ready"]),
  reviewing: new Set(["verified", "implementing"]),
  verified: new Set(["accepted", "implementing"]),
  accepted: new Set(["release_approved", "requirement_ready"]),
  release_approved: new Set(["deploying", "implementing"]),
  deploying: new Set(["operating", "implementing"]),
  operating: new Set(["outcome_validated", "implementing"]),
  outcome_validated: new Set(["requirement_ready"]),
};

const requiredEvidence: Partial<Record<LifecycleState, Set<string>>> = {
  intent_confirmed: new Set(["intent-approval"]),
  requirement_ready: new Set(["requirement-spec"]),
  design_approved: new Set(["technical-design", "design-approval"]),
  reviewing: new Set(["change-request"]),
  verified: new Set(["code-review", "quality-verification"]),
  accepted: new Set(["acceptance-approval"]),
  release_approved: new Set(["release-plan", "release-approval"]),
  operating: new Set(["deployment-record"]),
  outcome_validated: new Set(["outcome-validation"]),
};

export function missingEvidence(change: Change, target: LifecycleState): string[] {
  const present = new Set(change.evidence.filter((item) => ["passed", "approved", "valid"].includes(item.status.toLowerCase())).map((item) => item.kind));
  return [...(requiredEvidence[target] ?? [])].filter((kind) => !present.has(kind));
}

export function transition(change: Change, target: LifecycleState, options: {
  actor: string; reason: string; evidenceIds?: string[]; enforceGates?: boolean;
}): Change {
  if (!allowed[change.state].has(target)) throw new TransitionError(`cannot transition from ${change.state} to ${target}`);
  const missing = options.enforceGates === false ? [] : missingEvidence(change, target);
  if (missing.length) throw new TransitionError(`transition gate is missing evidence: ${missing.sort().join(", ")}`);
  const previous = change.state;
  change.state = target;
  change.updatedAt = utcNow();
  change.history.push({ fromState: previous, toState: target, actor: options.actor, reason: options.reason, at: utcNow(), evidenceIds: options.evidenceIds ?? [] });
  return change;
}
